### PR TITLE
feat: add resync process to handle polaris server failure

### LINKED ISSUE #66: Add polaris server failure detection and instance resyncing to increase robustness
**What is the feature you want to add?**
polaris server failure detection and instance resyncing

**Why do you want to add this feature?**
1. There is not method to detect the server failure currently.
2. When server fails and restarts, the data of instance will not sync uitl resource changing.

**How to implement this feature?**
1. Use probe request to check polaris server's health
2. Once server failure is detected, add resync task to queque. When the server is up, start the task in the queue.

**Additional context**

### PR DESCRIPTION
**Pull request by lhiamgeek**
**Please provide issue(s) of this PR:**
Fixes #66 

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Docs
- [ ] Inject Sidecar
- [ ] Installation
- [x] Performance and Scalability
- [ ] Test and Release

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any user-facing changes. This may include API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by lhiamgeek** [comment]
design:
1. add one more resync local cache to store services need to sync
2. periodically check health of poplaris server. If server fails, clear the origin local cache.
3.  periodically compare the resync local cache and origin cache and put any differences into worker queue.

**Comment by lhiamgeek** [comment]
+---------------------------+
                                                                                |                           |
                                                                                |                           |
  +--------------------+                                                        |                           |
  |                    |                                                        |                           |
  |                    |                                                        |                           |
  |    health checker  +------------------------+-------------------------------+       Polaris server      |
  |                    |                        |                               |                           |
  |                    |                        |                               |                           |
  +--------------------+                        |                               |                           |
                                      +---------v------------+                  |                           |
                                      |                      |                  |                           |
                                      |   failure detected   |                  +---------------------------+
                                      |                      |
                                      +---------+------------+
                                                |
                                                |
                                                |
                                                |clear
                                                |
                                                |
                                                |
                                                |
                                     +----------v-----------+
                                     |                      |
                                     |                      |
                           XXXX      |                      |
                          XX         |     local  cache     |
                          X          |                      +------------+
                          X          |                      |            |
+--------------------+    X          |                      |            |
|                    |    X          |                      |            |             +-------------------------------+
|                    |   XX          +----------------------+            |             |                               |
|   resync worker    | XXX                                               |    diff     |                               |
|                    |   XX                                              +------------>+       worker queue            |
|                    |    X                                              ^             |                               |
+--------------------+    X                                              |             |                               |
                          X         +------------------------+           |             +-------------------------------+
                          X         |                        |           |
                          X         |                        |           |
                          XX        |                        |           |
                           XXXX     |     resync cache       +-----------+
                                    |                        |
                                    |                        |
                                    |                        |
                                    |                        |
                                    +------------------------+

File: pkg/cache/service_cache.go [hunk 1]
```diff
@@ -55,3 +55,16 @@ func (csm *CachedServiceMap) Load(key string) (value *v1.Service, ok bool) {
...
+
+// Clear remove all elements in cache
+func (csm *CachedServiceMap) Clear() {
+	csm.sm.Range(func(key interface{}, value interface{}) bool {
+		csm.sm.Delete(key)
+		return true
+	})
+}
+
+// Range execute f for each element of cache
```
Comment (chuntaojun) [inline pkg/cache/service_cache.go:67]: 建议value是一个明确的类型，在方法内部强壮，减少上层使用的理解负担

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -217,6 +229,12 @@ func (p *PolarisController) Run(workers int, stopCh <-chan struct{}) {
 	// 定时任务
 	go p.MetricTracker(stopCh)
 
+	// 定时对账
+	go wait.Until(p.resyncWorker, time.Second*30, stopCh)
```
Comment (chuntaojun) [inline pkg/controller/controller.go:233]: 建议这里可以通过配置进行控制

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -217,6 +229,12 @@ func (p *PolarisController) Run(workers int, stopCh <-chan struct{}) {
...
+
+	// 定时健康探测
+	go wait.Until(p.checkHealth, time.Second, stopCh)
```
Comment (chuntaojun) [inline pkg/controller/controller.go:236]: 同上

Comment (lhiamgeek) [reply pkg/controller/controller.go:236]: 好的,这里后续我会提一个pr改成config

File: pkg/cache/service_cache.go (hunk 1 above)
Comment (lhiamgeek) [reply pkg/cache/service_cache.go:67]: 好的👍

**Comment by lhiamgeek** [comment]
/// +------------------------------------------------------------------------------------------------+
/// |                                                                                                |
/// |                                                                                                |
/// |                                                                                                |
/// |                                                                                                |
/// |                                                                                                |
/// |         +-----------------+                                        +-----------------+         |
/// |         |                 |                                        |                 |         |
/// |         | health checker  +------------------+---------------------+ polaris server  |         |
/// |         |                 |                  |                     |                 |         |
/// |         +-----------------+                  |                     +-----------------+         |
/// |                                              |                                                 |
/// |                                              |failure detected                                 |
/// |                                              |clear local cache                                |
/// +----------------------------------------------+-------------------------------------------------+
///                                        
       |
///                                                |
/// +----------------------------------------------+-------------------------------------------------+
/// |                                      
       |                                                 |
/// |                                          +---v--------+
                                       |
/// |                                          |            |
                                       |
/// |                                +---------+ local cache+-----+                                  |
/// |                                |         |            |
    |                                  |
/// |                                |         +------------+
    |                                  |
/// |                                |                            |                                  |
/// |        +---------------------+ |                            |                                  |
/// |        |                     | |                            |                +--------------+  |
/// |        |   resync worker     +-+                            |     diff       |              |  |
/// |        |                     | |                            <--------------->| worker queue |  |
/// |        +---------------------+ |                            |                |              |  |
/// |                                |                            |                +--------------+  |
/// |                                |                            |                                  |
/// |                                |                            |                                  |
/// |                                |         +-------------+    |                                  |
/// |                                |         |             |    |                                  |
/// |                                +---------+ resync cache+----+                                  |
/// |                                          |             |                                       |
/// |                                          +-------------+                                       |
/// |                                                                                                |
/// +------------------------------------------------------------------------------------------------+

**Comment by lhiamgeek** [comment]
https://asciiflow.com/#/share/eJztVt1qgzAUfpWQ615tjFGfJTchHlGW6ZbqVimF0eteStmzjD6NTzLbTlZr1CQ9boU2nAtjwvfnMbqgMX8G6sWZlBMqeQ6KenTB6JxRb%2FpwP2E0r67uHqfVVQrztJowSspie4XFWLxzTpDHDfnPkZ1bwILLkaCpUgdsoUF774QhBC7TkIgQxBOoM6SjFHlJJFfRjMxAvf3IccqkKyizVLC6o0fF%2Bf1hNCwaxgU74JHMFBAfUhAp%2BJjYQgJXRCaCSyJ41aDD2FdYO%2BcDgzHVTMpwHHb%2Fu8NLq0MmNpmbBt5O3mzvb22%2BeoWrYTgH%2FtXJFJOnLNa9tTo%2BILSukVQcG2wu1X5H4SrK4sOgPsfToAewYUF56%2FA1dcLYM1mbaSlRMMtjQd4Ttf8LqzHXA7bqZT8Kgi55Wr4u5%2F1sG8vHtj%2BOalOvGWSgUYPSG0Yp9d060WUBNczk3Bt4KrSbLo7FMqixNOjW8JiOqv0929YnQccHDUNDy1vnHI3F7qki87uN8vb3qytGl3T5DTOVXaA%3D)

File: cmd/polaris-controller/app/options/polaris.go [hunk 4]
```diff
@@ -36,6 +36,8 @@ type PolarisControllerConfiguration struct {
 	MinAccountingPeriod    metav1.Duration
 	SyncMode               string
 	SidecarMode            string
+	HealthCheckDuration    int
```
Comment (chuntaojun) [inline cmd/polaris-controller/app/options/polaris.go:39]: 要不用 time.ParseDuration(string) 的方式来设置？这样子会不会使用起来好一点？

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -957,3 +989,50 @@ func (p *PolarisController) MetricTracker(stopCh <-chan struct{}) {
...
+func (p *PolarisController) resyncWorker() {
+	if !p.isPolarisServerHealthy.Load() {
+		log.Info("Resync: Polaris server failed, not sync")
+		return
+	}
+
+	p.resyncServiceCache.Range(func(key string, value *v1.Service) bool {
+		v, ok := p.serviceCache.Load(util.GetOriginKeyWithResyncQueueKey(key))
+		if !ok {
+			p.onServiceAdd(value)
```
Comment (chuntaojun) [inline pkg/controller/controller.go:1003]: p.onServiceAdd => p.enqueueService(key, service, "Add")

**Review by chuntaojun — APPROVED** [review]

**Pull request merged**